File: frontend/lib/models/transaction.dart
/// Transaction model
enum TransactionType { INCOME, EXPENSE, TRANSFER }
enum TransactionSource { MANUAL, SMS, OCR, BANK_SYNC }
enum RecurringFrequency { DAILY, WEEKLY, MONTHLY, YEARLY }
enum RichDadCategory { 
  ACTIVE_INCOME, 
  PASSIVE_INCOME, 
  ASSET_EXPENSE, 
  LIABILITY_EXPENSE, 
  NECESSITY 
}

class Transaction {
  final String id;
  final String userId;
  final TransactionType type;
  final String category;
  final String? subCategory;
  final double amount;
  final String currency;
  final String? description;
  final String? merchantName;
  final TransactionSource source;
  final String? accountIdentifier;
  final DateTime transactionDate;
  final bool isRecurring;
  final RecurringFrequency? recurringFrequency;
  final RichDadCategory? richDadCategory;
  final DateTime createdAt;
  final DateTime updatedAt;

  Transaction({
    required this.id,
    required this.userId,
    required this.type,
    required this.category,
    this.subCategory,
    required this.amount,
    this.currency = 'INR',
    this.description,
    this.merchantName,
    this.source = TransactionSource.MANUAL,
    this.accountIdentifier,
    required this.transactionDate,
    this.isRecurring = false,
    this.recurringFrequency,
    this.richDadCategory,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'],
      userId: json['user_id'],
      type: TransactionType.values.byName(json['type']),
      category: json['category'],
      subCategory: json['sub_category'],
      amount: (json['amount'] as num).toDouble(),
      currency: json['currency'] ?? 'INR',
      description: json['description'],
      merchantName: json['merchant_name'],
      source: TransactionSource.values.byName(json['source']),
      accountIdentifier: json['account_identifier'],
      transactionDate: DateTime.parse(json['transaction_date']),
      isRecurring: json['is_recurring'] ?? false,
      recurringFrequency: json['recurring_frequency'] != null
          ? RecurringFrequency.values.byName(json['recurring_frequency'])
          : null,
      richDadCategory: json['rich_dad_category'] != null
          ? RichDadCategory.values.byName(json['rich_dad_category'])
          : null,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'type': type.name,
      'category': category,
      'sub_category': subCategory,
      'amount': amount,
      'currency': currency,
      'description': description,
      'merchant_name': merchantName,
      'source': source.name,
      'account_identifier': accountIdentifier,
      'transaction_date': transactionDate.toIso8601String().split('T')[0],
      'is_recurring': isRecurring,
      'recurring_frequency': recurringFrequency?.name,
      'rich_dad_category': richDadCategory?.name,
    };
  }
}

File: backend/app/models/transaction.py
"""Transaction model."""
import uuid
from datetime import datetime, date
from decimal import Decimal
from sqlalchemy import Column, String, DateTime, Date, Boolean, Enum, Numeric, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
import enum
from app.database import Base


class TransactionType(str, enum.Enum):
    """Transaction type enumeration."""
    INCOME = "INCOME"
    EXPENSE = "EXPENSE"
    TRANSFER = "TRANSFER"


class TransactionSource(str, enum.Enum):
    """Transaction source enumeration."""
    MANUAL = "MANUAL"
    SMS = "SMS"
    OCR = "OCR"
    BANK_SYNC = "BANK_SYNC"


class RecurringFrequency(str, enum.Enum):
    """Recurring frequency enumeration."""
    DAILY = "DAILY"
    WEEKLY = "WEEKLY"
    MONTHLY = "MONTHLY"
    YEARLY = "YEARLY"


class RichDadCategory(str, enum.Enum):
    """Rich Dad category enumeration."""
    ACTIVE_INCOME = "ACTIVE_INCOME"
    PASSIVE_INCOME = "PASSIVE_INCOME"
    ASSET_EXPENSE = "ASSET_EXPENSE"
    LIABILITY_EXPENSE = "LIABILITY_EXPENSE"
    NECESSITY = "NECESSITY"


class Transaction(Base):
    """Transaction model for income and expenses."""
    
    __tablename__ = "transactions"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    type = Column(Enum(TransactionType), nullable=False)
    category = Column(String, nullable=False)
    sub_category = Column(String, nullable=True)
    amount = Column(Numeric(15, 2), nullable=False)
    currency = Column(String, default="INR", nullable=False)
    description = Column(String, nullable=True)
    merchant_name = Column(String, nullable=True)
    source = Column(Enum(TransactionSource), default=TransactionSource.MANUAL, nullable=False)
    account_identifier = Column(String, nullable=True)  # Last 4 digits
    transaction_date = Column(Date, default=date.today, nullable=False)
    is_recurring = Column(Boolean, default=False, nullable=False)
    recurring_frequency = Column(Enum(RecurringFrequency), nullable=True)
    rich_dad_category = Column(Enum(RichDadCategory), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    user = relationship("User", back_populates="transactions")
